### PR TITLE
Upgrade YesSql to 4.0.0-beta-0003

### PR DESCRIPTION
**Pull request by MikeAlhayek**
@sebastienros @jtkech this PR uses the new `Async` methods provided by YesSql.

Here is a summary of all the changes

- Update YesSql to 4.0.0-beta-0002
- Use all the new asynchronous methods in YesSql
- Use `await` when creating a connection to dispose asynchronously.
- `BeginTransaction` asynchronously.
- Update the readme file

### CODE REVIEW

**Comment by MikeAlhayek** [comment]
@agriffard @Piedone @hishamco if you guys have time and don't mind, can you please test out this PR? @sebastienros is going to be doing some benchmarks to see the impact of it.

File: src/OrchardCore.Build/Dependencies.props
Comment (Piedone) [inline src/OrchardCore.Build/Dependencies.props:64]: Shouldn't the "Found version-specific or distribution-specific runtime identifier(s): alpine-arm, alpine-arm64, alpine-x64. Affected libraries: SQLitePCLRaw.lib.e_sqlite3. In .NET 8.0 and higher, assets for version-specific and distribution-specific runtime identifiers will not be found by default." build warnings be gone here too? With 3.5.0 they are.

File: src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs:69]: How about adding `[Obsolete]` on these sync-over-async methods?

File: src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs
Comment (Piedone) [inline src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs:75]: Why are these asyncs needed? It seems to me that none of them actually do any async work.

File: src/docs/reference/modules/Migrations/README.md
Comment (Piedone) [inline src/docs/reference/modules/Migrations/README.md:43]:
```suggestion
        public async Task<int> UpdateFrom1Async()
```

File: src/docs/resources/libraries/README.md
Comment (Piedone) [inline src/docs/resources/libraries/README.md:44]: We should probably wait for a release version before merging this, no?

**Review by Piedone — CHANGES_REQUESTED** [review]
image

nice

Functional tests look good: https://github.com/OrchardCMS/OrchardCore/actions/runs/6898591386

File: src/OrchardCore.Build/Dependencies.props
Comment (MikeAlhayek) [reply src/OrchardCore.Build/Dependencies.props:64]: The updates are not part of 0001.  I'll bump it up to 2 so we can remove the warning

File: src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs:69]: The interface already has the attribute

File: src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs
Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs:75]: We added these methods to  ISchemaBuiler interface in yessql. So we need to implement them here

File: src/docs/reference/modules/Migrations/README.md
Comment (MikeAlhayek) [reply src/docs/reference/modules/Migrations/README.md:43]: We would have to also update the calls to async if we change the signature. I'll update that in the AM

File: src/docs/resources/libraries/README.md
Comment (MikeAlhayek) [reply src/docs/resources/libraries/README.md:44]: We'll have a release before release OC. We need to do testing to ensure everything is good. @sebastienros is going to be doing benchmark too to understand any performance changes

File: src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs
Comment (Piedone) [reply src/OrchardCore/OrchardCore.Data.YesSql/Removing/ShellDbTablesInfo.cs:75]: Ah yeah, I missed that there's an interface from YesSql.

File: src/docs/reference/modules/Migrations/README.md
Comment (jtkech) [reply src/docs/reference/modules/Migrations/README.md:43]: We already can use both `UpdateFromN()` and `UpdateFromNAsync()`.

Comment (Piedone) [reply src/docs/reference/modules/Migrations/README.md:43]: I don't really understand; the caller is OC here, and OC supports *Async update methods. You changed the sample to use the async `SchemaBuilder` methods, so the method needs to be async too, because the current sample won't compile.

File: src/docs/resources/libraries/README.md
Comment (Piedone) [reply src/docs/resources/libraries/README.md:44]: And do you want to merge this before that, and then later update to 4.0.0 proper? I'm OK with it, just want to understand.

File: src/docs/reference/modules/Migrations/README.md
Comment (jtkech) [reply src/docs/reference/modules/Migrations/README.md:43]: Yes should use the async method, the automatic data migration already detect if it has to call a sync or async method

Comment (MikeAlhayek) [reply src/docs/reference/modules/Migrations/README.md:43]: You should using Async if your migration step does asynchronous call otherwise no need. In this PR, we are using all async methods to alter/create tables, so in this case we using the Async methods.  In OC both UpdateFrom1 and UpdateFrom1Async handle the same step. I think we first check if Async method exists and call it so it wins if both are defined in the same file

Comment (jtkech) [reply src/docs/reference/modules/Migrations/README.md:43]: No problem but now in the sample async call are done

File: src/docs/resources/libraries/README.md
Comment (MikeAlhayek) [reply src/docs/resources/libraries/README.md:44]: @Piedone 
Yes. We just need people to test it out. We'll release yessql 4 when @jtkech complete the PR that removed Newtonsoft library

File: src/docs/reference/modules/Migrations/README.md
Comment (jtkech) [reply src/docs/reference/modules/Migrations/README.md:43]: The following provided as a sample is wrong ;)

        public int UpdateFrom1()
        {
            await SchemaBuilder.CreateMapIndexTableAsync<MemberIndex>(table =>
            {

File: src/OrchardCore.Build/Dependencies.props
Comment (MikeAlhayek) [reply src/OrchardCore.Build/Dependencies.props]: I release and updated 0002. The warning should go away

**Review by Piedone — APPROVED** [review]

**Review by hishamco — APPROVED** [review]

**Comment by hishamco** [comment]
@MikeAlhayek is that mean that we can't use `UpdateFromX()` anymore?

**Comment by MikeAlhayek** [comment]
@hishamco 

> @MikeAlhayek is that mean that we can't use UpdateFromX() anymore?

You can use Both `UpdateFromX()` or `UpdateFromXAsync()`. This is an OC thing and it did not change. However, all methods in YesSql are now asynchronous and should be used over the non-async methods. Since we are switching to using Async methods in the migrations, we have to use `UpdateFromXAsync()`. But, if you have a migration step that does not do any asynchronous work, then use  `UpdateFromX()`.

**Comment by hishamco** [comment]
> This is an OC thing and it did not change

I knew, I thought there's no need for that while we move all the migrations to the async version

Let's merge then

**Comment by MikeAlhayek** [comment]
> Let's merge then

@sebastienros wants to do bench marking first to analyze the performance change this change bring. Once he approves this PR and let us know his finding, we'll merge it.

**Comment by sebastienros** [comment]
Similar RPS

```
| application                             | main          | branch        |         |
| --------------------------------------- | ------------- | ------------- | ------- |
| Max CPU Usage (%)                       |            98 |            99 |  +1.02% |
| Max Cores usage (%)                     |         1,179 |         1,186 |  +0.59% |
| Max Working Set (MB)                    |           874 |           815 |  -6.75% |
| Max Private Memory (MB)                 |         1,047 |           922 | -11.94% |
| Build Time (ms)                         |        26,941 |        49,870 | +85.11% |
| Start Time (ms)                         |           222 |           233 |  +4.95% |
| Published Size (KB)                     |       320,534 |       320,627 |  +0.03% |
| Symbols Size (KB)                       |         9,657 |         9,666 |  +0.09% |
| .NET Core SDK Version                   |       8.0.100 |       8.0.100 |         |
| Max CPU Usage (%)                       |            98 |            98 |  -0.24% |
| Max Working Set (MB)                    |           916 |           884 |  -3.51% |
| Max GC Heap Size (MB)                   |           567 |           434 | -23.44% |
| Size of committed memory by the GC (MB) |           677 |           641 |  -5.31% |
| Max Number of Gen 0 GCs / sec           |         31.00 |         31.00 |   0.00% |
| Max Number of Gen 1 GCs / sec           |          1.00 |          1.00 |   0.00% |
| Max Number of Gen 2 GCs / sec           |          1.00 |          1.00 |   0.00% |
| Max Gen 0 GC Budget (MB)                |           653 |           640 |  -1.99% |
| Max Time in GC (%)                      |          7.00 |          7.00 |   0.00% |
| Max Gen 0 Size (B)                      |     9,821,184 |    13,091,552 | +33.30% |
| Max Gen 1 Size (B)                      |    18,343,128 |    18,372,616 |  +0.16% |
| Max Gen 2 Size (B)                      |    32,750,536 |    32,728,688 |  -0.07% |
| Max LOH Size (B)                        |       524,664 |       384,456 | -26.72% |
| Max POH Size (B)                        |       768,296 |       894,760 | +16.46% |
| Max Allocation Rate (B/sec)             | 2,817,927,696 | 2,752,733,184 |  -2.31% |
| Max GC Heap Fragmentation (%)           |        1,592% |        1,764% | +10.80% |
| # of Assemblies Loaded                  |           389 |           389 |   0.00% |
| Max Exceptions (#/s)                    |            21 |            17 | -19.05% |
| Max Lock Contention (#/s)               |            23 |            37 | +60.87% |
| Max ThreadPool Threads Count            |            32 |            32 |   0.00% |
| Max ThreadPool Queue Length             |            26 |            20 | -23.08% |
| Max ThreadPool Items (#/s)              |        38,039 |        37,645 |  -1.04% |
| Max Active Timers                       |             7 |             7 |   0.00% |
| IL Jitted (B)                           |     2,703,712 |     2,735,906 |  +1.19% |
| Methods Jitted                          |        34,283 |        34,547 |  +0.77% |


| load                    | main    | branch  |          |
| ----------------------- | ------- | ------- | -------- |
| Max CPU Usage (%)       |       1 |       1 |    0.00% |
| Max Cores usage (%)     |      16 |      15 |   -6.25% |
| Max Working Set (MB)    |      52 |      61 |  +17.31% |
| Max Private Memory (MB) |     144 |     148 |   +2.78% |
| Build Time (ms)         |   2,190 |   5,650 | +157.99% |
| Start Time (ms)         |      40 |      37 |   -7.50% |
| Published Size (KB)     |  73,904 |  73,904 |    0.00% |
| Symbols Size (KB)       |       0 |       0 |          |
| .NET Core SDK Version   | 7.0.404 | 7.0.404 |          |
| First Request (ms)      |   2,205 |   2,210 |   +0.23% |
| Requests                | 191,344 | 188,782 |   -1.34% |
| Bad responses           |       0 |       0 |          |
| Latency 50th (ms)       |    2.04 |    2.10 |   +3.04% |
| Latency 75th (ms)       |    2.71 |    2.77 |   +2.17% |
| Latency 90th (ms)       |    3.73 |    3.83 |   +2.52% |
| Latency 95th (ms)       |    4.85 |    4.82 |   -0.62% |
| Latency 99th (ms)       |   12.61 |   12.14 |   -3.73% |
| Mean latency (ms)       |    2.51 |    2.54 |   +1.36% |
| Max latency (ms)        |   58.76 |   49.50 |  -15.77% |
| Requests/sec            |  12,757 |  12,586 |   -1.34% |
| Requests/sec (max)      |  14,292 |  14,284 |   -0.06% |
| Read throughput (MB/s)  |   74.86 |   73.81 |   -1.40% |
```

**Comment by sebastienros** [comment]
Crank command

```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/orchard.benchmarks.yml --scenario about-postgresql --profile aspnet-perf-lin --application.framework net8.0 --application.options.collectCounters true --application.source.branchOrCommit ma/upgrade-yessql-4b1 --json c:\temp\oc2.json --chart
```

**Pull request merged**
